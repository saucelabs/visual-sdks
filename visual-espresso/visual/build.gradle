plugins {
    alias libs.plugins.android.library
    alias libs.plugins.apollographql
    id 'maven-publish'
}

android {
    namespace 'com.saucelabs.visual'
    compileSdk 34

    defaultConfig {
        minSdk 21
        versionCode 1
        versionName "0.0.0"
        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
        consumerProguardFiles "consumer-rules.pro"

        buildConfigField "String", "SAUCE_USERNAME", "\"${getProperty("SAUCE_USERNAME", "")}\""
        buildConfigField "String", "SAUCE_ACCESS_KEY", "\"${getProperty("SAUCE_ACCESS_KEY", "")}\""
        buildConfigField "String", "VERSION_NAME", "\"${android.defaultConfig.versionName}\""
    }

    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
        }
    }
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_11
        targetCompatibility JavaVersion.VERSION_11
    }
    publishing {
        singleVariant("release") {
            withSourcesJar()
            withJavadocJar()
        }
    }
    buildFeatures {
        buildConfig = true
    }
}

String getProperty(String name, String defaultValue) {
    if (rootProject.file("./local.properties").exists()) {
        def properties = new Properties()
        properties.load(rootProject.file("./local.properties").newDataInputStream())
        if (properties.containsKey(name)) {
            return properties[name]
        }
    }
    if (System.getenv(name) != null && !System.getenv(name).isBlank()) {
        return System.getenv(name)
    }
    return defaultValue
}

dependencies {
    implementation libs.appcompat
    implementation libs.material
    implementation libs.junit
    implementation libs.apollo.runtime
    implementation libs.apollo.rx3.support
    implementation libs.espresso.core
    implementation libs.uiautomator
    implementation libs.jsoup
    androidTestImplementation libs.ext.junit
}

ext {
    scmUrl = 'git@github.com:saucelabs/visual-sdks.git'
    artifactName = 'visual-espresso'
}

publishing {
    publications {
        release(MavenPublication) {
            groupId = android.namespace
            artifactId = "${artifactName}"
            version = android.defaultConfig.versionName
            afterEvaluate {
                from components.release
            }
            pom {
                name = "${artifactName}"
                description = 'Sauce Visual Espresso SDK'
                url = 'https://github.com/saucelabs/visual-sdks'
                scm {
                    connection = "scm:git:${scmUrl}"
                    developerConnection = "scm:git:${scmUrl}"
                    url = "${scmUrl}"
                }
                licenses {
                    license {
                        name = 'Apache-2.0'
                        url = 'https://github.com/saucelabs/visual-sdks/blob/main/LICENSE'
                    }
                }
                developers {
                    developer {
                        name = 'Sauce Labs Visual Team'
                        email = 'visual.team@saucelabs.com'
                        organization = 'Sauce Labs'
                        organizationUrl = 'https://saucelabs.com/'
                    }
                }
            }
        }
    }
    repositories {
        mavenLocal()
    }
}

apollo {
    service("service") {
        packageName.set("com.saucelabs.visual.graphql")
        generateKotlinModels.set(false)
        warnOnDeprecatedUsages.set(false)
    }
}

tasks.register('bumpVersion') {
    doLast {
        def releaseType = project.hasProperty("releaseType") ? project.releaseType : null
        if (!releaseType) {
            throw new GradleException("Please provide a release type using -PreleaseType=<major|minor|patch>")
        }

        def currentVersionName = android.defaultConfig.versionName
        def newVersionName = calculateNewVersion(currentVersionName, releaseType)
        persistVersionName(newVersionName)

        println "$newVersionName"
    }
}

static def calculateNewVersion(currentVersion, releaseType) {
    def versionParts = currentVersion.split('\\.').collect { it.toInteger() }
    if (versionParts.size() != 3) {
        throw new GradleException("Invalid versionName format: $currentVersion. Expected format: MAJOR.MINOR.PATCH")
    }

    def (major, minor, patch) = versionParts
    switch (releaseType) {
        case "major":
            major++
            minor = 0
            patch = 0
            break
        case "minor":
            minor++
            patch = 0
            break
        case "patch":
            patch++
            break
        default:
            throw new GradleException("Unknown release type: $releaseType. Use 'major', 'minor', or 'patch'.")
    }

    return "$major.$minor.$patch"
}

def persistVersionName(newVersionName) {
    def buildFile = file("build.gradle")
    def buildFileContent = buildFile.text.replaceAll(/versionName\s+"[^"]+"/, "versionName \"$newVersionName\"")
    buildFile.text = buildFileContent
}
