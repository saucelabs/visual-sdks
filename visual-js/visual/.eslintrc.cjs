module.exports = {
  extends: ['eslint:recommended', 'plugin:@typescript-eslint/recommended'],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: './tsconfig.json',
    tsconfigRootDir: __dirname,
    sourceType: 'module',
  },
  plugins: ['@typescript-eslint'],
  extends: [
    'plugin:@typescript-eslint/recommended',
    'plugin:prettier/recommended',
  ],
  root: true,
  env: {
    node: true,
    jest: true,
  },
  ignorePatterns: ['.eslintrc.js', 'codegen.ts', 'jest.config.js', 'src/graphql/__generated__'],
  rules: {
    '@typescript-eslint/interface-name-prefix': 'off',
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/explicit-module-boundary-types': 'off',
    '@typescript-eslint/no-explicit-any': 'off',
    
    // Allow unused vars that start with _
    "@typescript-eslint/no-unused-vars": [
      "warn",
      {
        "argsIgnorePattern": "^_",
        "varsIgnorePattern": "^_",
        "caughtErrorsIgnorePattern": "^_"
      }
    ],
  },
};
