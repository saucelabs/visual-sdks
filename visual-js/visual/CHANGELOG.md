# @saucelabs/visual

## 0.11.0

### Minor Changes

- 0bd1fac: Removed deprecated functionality.

## 0.10.2

### Patch Changes

- e7abe58: expose nativeClipSelector for full page config

## 0.10.1

### Patch Changes

- 7a88f8d: add ignore region selectors for appium native fullpage

## 0.10.0

### Minor Changes

- f797342: Unify result checking behavior into utility
  Add automatic retry mechanism to file uploads for timeouts

## 0.9.0

### Minor Changes

- 8572e1b: scrollElement native fps

## 0.8.3

### Patch Changes

- 5991881: Fill out statuses in sauceVisualResults for cypress
  Bump all dependencies to use new ignore region calculation system

## 0.8.2

### Patch Changes

- ff9cdb2: Rebase Storybook integration onto Playwright

## 0.8.1

### Patch Changes

- 7b1c3cc: Add playwright support

## 0.8.0

### Minor Changes

- 1399b9c: Debugging info when upload is reaching timeout

## 0.7.3

### Patch Changes

- 8a64e27: Update / regen types + update Cypress snapshot upload behavior

## 0.7.2

### Patch Changes

- 3f1e241: Enable merge baselines command

## 0.7.1

### Patch Changes

- d769340: Add command to merge baselines

## 0.7.0

### Minor Changes

- 4bafb87: Change default diffingMethod to BALANCED

## 0.6.6

### Patch Changes

- 58c9211: replace deprecated field in buildStatus query

## 0.6.5

### Patch Changes

- a1f4373: Update types

## 0.6.4

### Patch Changes

- c5c1ee6: Priotize custom id > build id when checking status

## 0.6.3

### Patch Changes

- f15ec66: Type UUID codegen, add return type from sauceVisualCheck

## 0.6.2

### Patch Changes

- 312a7d7: Add WebElement support for GraphQL API, update shared types

## 0.6.1

### Patch Changes

- 926996f: Add Skip support for SAUCE_VISUAL_SKIP

## 0.6.0

### Minor Changes

- c94072c: Remove deprecated API calls webwebdriverSession and webdriverSessionFromArchive

## 0.5.0

### Minor Changes

- 787387a: deprecate some full page options

## 0.4.0

### Minor Changes

- 0384c59: Bump minor after GH migration

### Patch Changes

- ed72709: migrate source to github

## 0.4.0

### Minor Changes

- move to github
