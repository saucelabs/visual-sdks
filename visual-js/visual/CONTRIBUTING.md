
# Sauce Labs Visual client library


## Installation

```sh
yarn install
```

## Regenerate bindings

```sh
yarn copy-schema
yarn gen
yarn build
```

## Test

```sh
yarn test
```
