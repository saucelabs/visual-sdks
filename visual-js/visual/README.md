
# Sauce Labs Visual client library

This package provides a client and a CLI tool to interact with Sauce Labs Visual API.


**Note:** There are integrations available for [Cypress](https://www.npmjs.com/package/@saucelabs/cypress-visual-plugin) and [WebdriverIO](https://www.npmjs.com/package/@saucelabs/wdio-sauce-visual-service).


## Installation

```sh
npm install --save @saucelabs/visual
```
