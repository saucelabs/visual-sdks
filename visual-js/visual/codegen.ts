import type { CodegenConfig } from '@graphql-codegen/cli';

const username = process.env.SAUCE_USERNAME;
const accessKey = process.env.SAUCE_ACCESS_KEY;

const codegenConfig: CodegenConfig = {
  schema: process.env.VISUAL_API_FOR_CODEGEN || `https://${username}:${accessKey}@api.us-west-1.saucelabs.com/v1/visual/graphql`,
  documents: [
    'src/**/*.ts',
    'src/**/*.graphql',
    '!src/**/__generated__/**/*',
  ],
  emitLegacyCommonJSImports: false,
  config: {
    namingConvention: {
      typeNames: 'change-case-all#pascalCase',
      transformUnderscore: true,
    },
    avoidOptionals: {
      field: true,
      inputValue: false,
      object: true,
      defaultValue: true,
    },
    scalars: {
      WebdriverSessionBlob: "string",
      WebdriverElementID: "string",
      Datetime: "Date | string | number",
      UUID: "string",
    }
  },
  generates: {
    './src/graphql/__generated__/': {
      preset: 'client',
      plugins: [],
    },
  },
};

export default codegenConfig;
