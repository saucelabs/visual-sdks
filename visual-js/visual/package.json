{
  "name": "@saucelabs/visual",
  "description": "JS client bindings for Sauce Labs Visual",
  "version": "0.12.0",
  "main": "build/index.cjs",
  "module": "build/index.js",
  "types": "build/index.d.ts",
  "license": "MIT",
  "files": [
    "build",
    "README.md"
  ],
  "type": "module",
  "engines": {
    "node": "^16.13 || >=18"
  },
  "bin": "build/cli.js",
  "typeScriptVersion": "3.8.3",
  "keywords": [
    "webdriver",
    "wdio",
    "wdio-service",
    "saucelabs",
    "visual"
  ],
  "exports": {
    ".": {
      "require": "./build/index.cjs",
      "import": "./build/index.js",
      "types": "./build/index.d.ts"
    },
    "./package.json": "./package.json"
  },
  "scripts": {
    "build": "tsup",
    "watch": "tsc-watch --declaration -p .",
    "lint": "eslint \"{src,apps,libs,test}/**/*.ts\"",
    "gen": "npm run gen:graphql",
    "gen:graphql": "graphql-codegen --config codegen.ts",
    "pretest": "npm run gen",
    "test": "node --experimental-vm-modules $(yarn bin jest) --collect-coverage"
  },
  "dependencies": {
    "@apollo/client": "^3.7.14",
    "@graphql-typed-document-node/core": "3.2.0",
    "arktype": "^1.0.29-alpha",
    "axios": "^1.7.4",
    "chalk": "^4.0.0",
    "cli-table3": "^0.6.3",
    "commander": "^12.0.0",
    "exponential-backoff": "^3.1.1",
    "proxy-agent": "^6.3.1"
  },
  "tsup": {
    "entry": [
      "./src/index.ts",
      "./src/cli.ts"
    ],
    "dts": true,
    "outDir": "./build",
    "format": [
      "cjs",
      "esm"
    ],
    "noExternal": [
      "@apollo/client"
    ]
  },
  "devDependencies": {
    "@graphql-codegen/cli": "3.3.1",
    "@graphql-codegen/client-preset": "^3.0.1",
    "@jest/globals": "^29.5.0",
    "@tsconfig/node18": "^2.0.0",
    "@types/argparse": "^2.0.14",
    "@types/node": "^18.13.0",
    "@types/node-fetch": "^2.6.4",
    "@types/selenium-webdriver": "^4.1.15",
    "@typescript-eslint/eslint-plugin": "^5.59.1",
    "@typescript-eslint/parser": "^5.59.1",
    "eslint": "^8.0.1",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-prettier": "^4.0.0",
    "graphql": "^15.0.0",
    "jest": "^29.5.0",
    "jest-junit-reporter": "^1.1.0",
    "prettier": "^2.8.8",
    "selenium-webdriver": "^4.10.0",
    "ts-jest": "^29.1.1",
    "ts-jest-resolver": "^2.0.1",
    "ts-node": "^10.9.1",
    "tsc-watch": "^6.0.4",
    "tsup": "^7.2.0",
    "typescript": "^5.0.4"
  }
}
