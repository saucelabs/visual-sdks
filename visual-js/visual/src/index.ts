export * from './utils.js';
export * from './common/api.js';
export * from './types.js';
export { regions, VisualApiRegion } from './common/regions.js';
export { displayStatusTable } from './common/table.js';
