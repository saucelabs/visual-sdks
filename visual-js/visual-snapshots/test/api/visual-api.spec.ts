import { BuildStatus, VisualApi } from "@saucelabs/visual";
import {
  CreateVisualSnapshotsParams,
  VisualSnapshotsApi,
} from "../../src/api/visual-snapshots-api.js";
import {
  PdfSnapshotUploader,
  UploadPdfSnapshotsParams,
} from "../../src/api/pdf-files-snapshot-uploader.js";

describe("VisualSnapshots", () => {
  describe("generateAndSendPdfFileSnapshots", () => {
    const consoleInfoSpy = jest
      .spyOn(console, "info")
      .mockImplementation(() => undefined);

    const uploadSnapshotsMock = jest
      .fn<
        ReturnType<PdfSnapshotUploader["uploadSnapshots"]>,
        Parameters<PdfSnapshotUploader["uploadSnapshots"]>
      >()
      .mockResolvedValue(Promise.resolve());

    const pdfFilesSnapshotUploaderMock: PdfSnapshotUploader = {
      uploadSnapshots: uploadSnapshotsMock,
    };

    const createBuildMock = jest.fn<
      ReturnType<VisualApi["createBuild"]>,
      Parameters<VisualApi["createBuild"]>
    >();
    const finishBuildMock = jest.fn<
      ReturnType<VisualApi["finishBuild"]>,
      Parameters<VisualApi["finishBuild"]>
    >();
    const buildStatusMock = jest.fn<
      ReturnType<VisualApi["buildStatus"]>,
      Parameters<VisualApi["buildStatus"]>
    >();
    const visualApiMock = {
      createBuild: createBuildMock,
      finishBuild: finishBuildMock,
      buildStatus: buildStatusMock,
    } as never as VisualApi;

    const visualSnapshots = new VisualSnapshotsApi(
      visualApiMock,
      pdfFilesSnapshotUploaderMock
    );

    const files = ["file1.pdf", "file2.pdf"];

    beforeEach(() => {
      createBuildMock.mockReset();
      createBuildMock.mockResolvedValue({
        id: "build-id",
        url: "http://build-url/build-id",
      } as never);

      uploadSnapshotsMock.mockReset();
      finishBuildMock.mockReset();
      buildStatusMock.mockReset();
      consoleInfoSpy.mockReset();
    });

    function assertBuild(params: CreateVisualSnapshotsParams) {
      if (!params.buildId) {
        expect(createBuildMock).toHaveBeenCalledWith({
          name: params.buildName,
          branch: params.branch,
          defaultBranch: params.defaultBranch,
          project: params.project,
          customId: params.customId,
        });
      } else {
        expect(createBuildMock).not.toHaveBeenCalled();
      }

      expect(finishBuildMock).toHaveBeenCalledWith({
        uuid: params.buildId ?? "build-id",
      });

      expect(buildStatusMock).toHaveBeenCalledWith(
        params.buildId ?? "build-id"
      );
    }

    async function assertSnapshot(
      pdfFilePaths: string[],
      params: CreateVisualSnapshotsParams
    ) {
      expect(uploadSnapshotsMock).toHaveBeenCalledWith({
        buildId: params.buildId ?? "build-id",
        pdfFilePaths,
        suiteName: params.suiteName,
        testNameFormat: params.testName,
        snapshotNameFormat: params.snapshotName,
      } satisfies UploadPdfSnapshotsParams);
    }

    describe("with params", () => {
      const params = {
        branch: "fake-branch",
        buildName: "fake-build-name",
        defaultBranch: "fake-default-branch",
        project: "fake-project",
        customId: "fake-custom-id",
        snapshotName: "custom-snapshot-name-{filename}-{page}",
        suiteName: "custom-suite-name",
        testName: "custom-test-name-{filename}",
      } satisfies CreateVisualSnapshotsParams;

      test("difffing unfinished", async () => {
        buildStatusMock.mockResolvedValueOnce({
          status: BuildStatus.Running,
          unapprovedCount: 2,
          errorCount: 0,
          url: "",
        });

        await visualSnapshots.generateAndSendPdfFileSnapshots(files, params);

        assertBuild(params);
        await assertSnapshot(files, params);

        expect(consoleInfoSpy.mock.calls).toMatchSnapshot();
      });

      test("difffing finished", async () => {
        buildStatusMock.mockResolvedValueOnce({
          status: BuildStatus.Approved,
          unapprovedCount: 0,
          errorCount: 0,
          url: "",
        });

        await visualSnapshots.generateAndSendPdfFileSnapshots(files, params);

        assertBuild(params);
        await assertSnapshot(files, params);

        expect(consoleInfoSpy.mock.calls).toMatchSnapshot();
      });
    });

    describe("with params and build-id", () => {
      const params = {
        branch: "fake-branch",
        buildName: "fake-build-name",
        defaultBranch: "fake-default-branch",
        project: "fake-project",
        customId: "fake-custom-id",
        buildId: "custom-build-id",
        snapshotName: "custom-snapshot-name-{filename}-{page}",
        suiteName: "custom-suite-name",
        testName: "custom-test-name-{filename}",
      } satisfies CreateVisualSnapshotsParams;

      test("difffing unfinished", async () => {
        buildStatusMock.mockResolvedValueOnce({
          status: BuildStatus.Running,
          unapprovedCount: 2,
          errorCount: 0,
          url: "",
        });

        await visualSnapshots.generateAndSendPdfFileSnapshots(files, params);

        assertBuild(params);
        await assertSnapshot(files, params);

        expect(consoleInfoSpy.mock.calls).toMatchSnapshot();
      });

      test("difffing finished", async () => {
        buildStatusMock.mockResolvedValueOnce({
          status: BuildStatus.Approved,
          unapprovedCount: 0,
          errorCount: 0,
          url: "",
        });

        await visualSnapshots.generateAndSendPdfFileSnapshots(files, params);

        assertBuild(params);
        await assertSnapshot(files, params);

        expect(consoleInfoSpy.mock.calls).toMatchSnapshot();
      });
    });

    test("without params", async () => {
      buildStatusMock.mockResolvedValueOnce({
        status: BuildStatus.Unapproved,
        unapprovedCount: 2,
        errorCount: 0,
        url: "",
      });

      const params = {} as CreateVisualSnapshotsParams;
      await visualSnapshots.generateAndSendPdfFileSnapshots(files, params);

      assertBuild(params);
      await assertSnapshot(files, params);

      expect(consoleInfoSpy.mock.calls).toMatchSnapshot();
    });
  });
});
