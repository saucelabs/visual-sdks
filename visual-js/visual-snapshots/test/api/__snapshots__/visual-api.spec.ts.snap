// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VisualSnapshots generateAndSendPdfFileSnapshots with params and build-id difffing finished 1`] = `
[
  [
    "Build custom-build-id finished.",
  ],
  [
    "Build custom-build-id finished (status=APPROVED, unapprovedCount=0, errorCount=0).",
  ],
]
`;

exports[`VisualSnapshots generateAndSendPdfFileSnapshots with params and build-id difffing unfinished 1`] = `
[
  [
    "Build custom-build-id finished.",
  ],
  [
    "Build custom-build-id finished but snapshots haven't been compared yet. Check the build status in a few moments.",
  ],
]
`;

exports[`VisualSnapshots generateAndSendPdfFileSnapshots with params difffing finished 1`] = `
[
  [
    "Build build-id created: http://build-url/build-id",
  ],
  [
    "Build build-id finished.",
  ],
  [
    "Build build-id finished (status=APPROVED, unapprovedCount=0, errorCount=0).",
  ],
]
`;

exports[`VisualSnapshots generateAndSendPdfFileSnapshots with params difffing unfinished 1`] = `
[
  [
    "Build build-id created: http://build-url/build-id",
  ],
  [
    "Build build-id finished.",
  ],
  [
    "Build build-id finished but snapshots haven't been compared yet. Check the build status in a few moments.",
  ],
]
`;

exports[`VisualSnapshots generateAndSendPdfFileSnapshots without params 1`] = `
[
  [
    "Build build-id created: http://build-url/build-id",
  ],
  [
    "Build build-id finished.",
  ],
  [
    "Build build-id finished (status=UNAPPROVED, unapprovedCount=2, errorCount=0).",
  ],
]
`;
